packages/api/src/router/post.ts

```diff
@@ -1,27 +1,72 @@
+import { TRPCError } from "@trpc/server";
 import { z } from "zod";
 
-import { createTRPCRouter, publicProcedure } from "../trpc";
+import { createTRPCRouter, protectedProcedure, publicProcedure } from "../trpc";
 
 export const postRouter = createTRPCRouter({
   all: publicProcedure.query(({ ctx }) => {
-    return ctx.prisma.post.findMany({ orderBy: { id: "desc" } });
+    return ctx.prisma.post.findMany({
+      include: { author: { select: { name: true, image: true } } },
+      orderBy: { id: "desc" },
+    });
   }),
+
   byId: publicProcedure
     .input(z.object({ id: z.string() }))
     .query(({ ctx, input }) => {
-      return ctx.prisma.post.findFirst({ where: { id: input.id } });
+      return ctx.prisma.post.findFirst({
+        include: { author: { select: { name: true, image: true } } },
+        where: { id: input.id },
+      });
     }),
-  create: publicProcedure
+
+  create: protectedProcedure
     .input(
       z.object({
         title: z.string().min(1),
         content: z.string().min(1),
       }),
     )
-    .mutation(({ ctx, input }) => {
-      return ctx.prisma.post.create({ data: input });
+    .mutation(async ({ ctx, input }) => {
+      function getNameFromUser() {
+        const meta = ctx.user.user_metadata;
+        if (typeof meta.name === "string") return meta.name;
+        if (typeof meta.full_name === "string") return meta.full_name;
+        if (typeof meta.user_name === "string") return meta.user_name;
+        return "[redacted]";
+      }
+
+      return ctx.prisma.post.create({
+        data: {
+          author: {
+            connectOrCreate: {
+              where: { id: ctx.user.id },
+              create: {
+                id: ctx.user.id,
+                email: ctx.user.email,
+                name: getNameFromUser(),
+                image: ctx.user.user_metadata.avatar_url as string | undefined,
```

Comment:
does every provider use the `avatar_url` field or should there be some `getImageFromUser` function like the name above to make sure this matches all possible fields?

Reply:
It's termed as `avatar_url` on most if not all of the major providers (e.g. Google, Github. etc). There may be a few providers which use a different name but it'd be hard to tell in advance - think it's fine to leave as is and if users are using a provider which doesn't use `avatar_url` as a field we can reconsider whether to adjust